## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [22]:
import seaborn as sns
import pandas as pd
import numpy as np


titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
429,1,3,male,32.0,0,0,8.0500,S,Third,man,True,E,Southampton,yes,True
159,0,3,male,NaN,8,2,69.5500,S,Third,man,True,NaN,Southampton,no,False
641,1,1,female,24.0,0,0,69.3000,C,First,woman,False,B,Cherbourg,yes,True
218,1,1,female,32.0,0,0,76.2917,C,First,woman,False,D,Cherbourg,yes,True
832,0,3,male,NaN,0,0,7.2292,C,Third,man,True,NaN,Cherbourg,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [23]:
not_nan=titanic_data.isnull().sum()
not_nan

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [ ]:

corrected_data=titanic_data.dropna(axis=1,thresh=titanic_data.shape[0]/2)
corrected_data=corrected_data.dropna(axis=0,thresh=corrected_data.shape[1]/2)
corrected_data

445.5


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [30]:
for who in ["man", "woman", "child"]:
    mask = (corrected_data["age"].isnull()) & (corrected_data["who"] == who)
    median_val = corrected_data[
        (corrected_data["age"].notnull()) & (corrected_data["who"] == who)
    ]["age"].median()
    corrected_data.loc[mask, "age"] = (median_val.round())

corrected_data

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,30.0,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [31]:
corrected_data=corrected_data.dropna(thresh=len(corrected_data.columns)-1)
mask=corrected_data.isnull()

#проверка
mask1=mask.any(axis=1)
corrected_data[mask1]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone


### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [32]:
city=corrected_data["embark_town"].value_counts().index[0]
print(city)

Southampton


### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [ ]:
percent=round(corrected_data["alive"].value_counts()["yes"]/(len(corrected_data))*100,2)
print(f"Выживших {percent} %")


Выживших 38.25 %


340

### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [ ]:
mask_survived=corrected_data["alive"]=="yes"
survived_counts_cities=corrected_data[mask_survived]["embarked"].value_counts()
print(survived_counts_cities)

embarked
S    217
C     93
Q     30
Name: count, dtype: int64


### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [35]:
survived_counts_classes=corrected_data[mask_survived]["pclass"].value_counts()
all_by_classes=corrected_data["pclass"].value_counts()
percents=round(survived_counts_classes.div(all_by_classes, fill_value=0)*100,2)

print(percents)

pclass
1    62.62
2    47.28
3    24.24
Name: count, dtype: float64


### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [50]:
mask_rich=corrected_data["fare"]>=100
mask_rich_survived=(corrected_data["fare"]>=100) & (corrected_data["survived"]==1)
rich_percent_survived=round((len(corrected_data[mask_rich_survived]))/len(corrected_data[mask_rich])*100, 2)
print(f"Богатых выжило {rich_percent_survived} %")

Богатых выжило 73.58 %


### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [48]:
mask_alone_child=(corrected_data["who"]=="child") & (corrected_data["alone"])
num_alone_kids=len(corrected_data[mask_alone_child])
print(f"Детей путешествовало в одиночку: {num_alone_kids}")

Детей путешествовало в одиночку: 6


Какие выводы вы можете сделать о выживших пассажирах Титаника? 

Неудивительно, но выжило очень много богатых 💀💀💀
К тому же, судя по тому, что было 6 детей "без родителей", курс титаника проходил через остров Эпштейна